### ISSUE TITLE
Chebyshev polynomial smoother revisited

### ISSUE THREAD

**Issue by gamnipe** [ddemidov/amgcl#126]
This is a follow-up to issue #2 Chebyshev polynomial smoother.

I experienced severe problems with the present implementation of the polynomial smoother for my (certainly badly scaled but solvable) linear systems.

An alternative implementation for the Chebyshev smoother is described in Algorithm 1 in:
P. Ghysels, P. Kłosiewicz, and W. Vanroose. "Improving the arithmetic intensity of multigrid with the help of polynomial smoothers". Numer. Linear Algebra Appl. 2012;19:253-267. DOI: 10.1002/nla.1808. It does not require the setup/solution of a linear system for the coefficients. From my experience it could improve the performance/robustness significantly, especially in combination with an additional diagonal scaling during the smoothing process.

I just wondered 
1. whether there was a specific reason to choose the present implementation of the smoother and not an implementation as shown in the reference, and
2. whether others made a similar (or different) experience

**Comment by ddemidov** [comment]
Hello!

I have to say I have very little experience with the Chebyshev relaxation. There was no specific reason for the choice of the present algorithm: it was probably the first accessible description I found at the time.

Thank you for the suggestion! Do you have a readily available implementation of the variant you referenced? I would gladly accept a PR with the updated implementation. If not, I'll have a look at the paper myself when I have time.

**Comment by ddemidov** [comment]
4c92f34bbbbec34c260f243299eca2addc1c661a is an attempt to implement the algorithm from Ghysels et al. Unfortunately, it does not look like it improves things.
@gamnipe, do you see any flaws in my implementation?

Here is a simple test for a 3D Poisson problem on a 64^3 grid.
The 5-degree polynomial is used, and `lambda_max` is estimated with Gershgorin disk theorem in both cases: 

**Old implementation** converges in 6 iterations:
```
./solver_ref -n 64 -p precond.relax.type=chebyshev
Solver
======
Type:             BiCGStab
Unknowns:         262144
Memory footprint: 14.00 M

Preconditioner
==============
Number of levels:    3
Operator complexity: 1.61
Grid complexity:     1.13
Memory footprint:    106.14 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0       262144        1810432     70.52 M (62.20%)
    1        31868         955994     20.59 M (32.84%)
    2         2279         144245     15.04 M ( 4.96%)

Iterations: 6
Error:      8.1731e-10

[Profile:          1.533 s] (100.00%)
[ self:            0.002 s] (  0.15%)
[  assembling:     0.053 s] (  3.43%)
[  setup:          0.771 s] ( 50.28%)
[  solve:          0.707 s] ( 46.13%)
```

**New implementation** does not converge in 100 iterations:
```
./solver -n 64 -p precond.relax.type=chebyshev
Solver
======
Type:             BiCGStab
Unknowns:         262144
Memory footprint: 14.00 M

Preconditioner
==============
Number of levels:    3
Operator complexity: 1.61
Grid complexity:     1.13
Memory footprint:    103.90 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0       262144        1810432     68.52 M (62.20%)
    1        31868         955994     20.34 M (32.84%)
    2         2279         144245     15.04 M ( 4.96%)

Iterations: 100
Error:      0.798063

[Profile:          6.379 s] (100.00%)
[  assembling:     0.069 s] (  1.09%)
[  setup:          0.734 s] ( 11.50%)
[  solve:          5.573 s] ( 87.37%)

```

**Comment by ddemidov** [comment]
Ok, I can see the error.

**Comment by gamnipe** [comment]
I'm sorry, you're simply too fast

1) concerning your implementation: it looks like you're evaluating the residual Ax-b only once (upon calling apply_pre/apply_post) --- I think it needs to be called inside the loop over k (you want applications of A**k of the matrix for the polynomial) 

2) The answer to your previous question: I have an implementation (+scaling), but it was rather messy and I'm currently working on cleaning it up & add some comments. It works for my class of problems (and the architecture I use), but I'm not sure whether it will mess up performance for other applications

If you want and can do a pull request and you can take a look.

Cheers

Peter

**Comment by ddemidov** [comment]
> it looks like you're evaluating the residual Ax-b only once 

Yes, that was the reason. ef26eb064f010f75a1251bfd940c13c0ca8ea5d4 fixes that with this result:

```
./solver -n 64 -p precond.relax.type=chebyshev
Solver
======
Type:             BiCGStab
Unknowns:         262144
Memory footprint: 14.00 M

Preconditioner
==============
Number of levels:    3
Operator complexity: 1.61
Grid complexity:     1.13
Memory footprint:    106.14 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0       262144        1810432     70.52 M (62.20%)
    1        31868         955994     20.59 M (32.84%)
    2         2279         144245     15.04 M ( 4.96%)

Iterations: 6
Error:      8.17312e-10

[Profile:          1.517 s] (100.00%)
[ self:            0.003 s] (  0.18%)
[  assembling:     0.065 s] (  4.29%)
[  setup:          0.751 s] ( 49.46%)
[  solve:          0.699 s] ( 46.07%)
```

Now it looks like (judging by the final residual) the old and the new implementations are identical (although I prefer the simplicity of the new one).

**Comment by ddemidov** [comment]
So could the scaling be the main reason your version works better with your problems? What do you mean by scaling exactly?
